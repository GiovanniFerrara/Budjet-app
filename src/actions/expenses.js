import uuid from 'uuid';
import db from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseObj = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseObj;
    const expense = { description, note, amount, createdAt }
    return db.ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(addExpense({
          ...expense,
          id: ref.key
        }))
      })
      .catch(e => console.log(e))
  }
}


// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
