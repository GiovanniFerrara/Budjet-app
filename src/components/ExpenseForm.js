import React, { Component } from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'

const now = moment();
console.log(now.format("Do - MMM - YYYY"))

class ExpenseForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
        description: props.expense?  props.expense.description : "",
        createdAt: props.expense?  moment(props.expense.createdAt) : moment(),
        amount: props.expense?  (props.expense.amount/100).toString() : 0,
        note: props.expense? props.expense.note : "",
        calendarFocus: false,
        error: "",
    }
  }
  handleSubmit = (e) => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount ){
      this.setState(()=>({
        error : "Please enter description and amount"
      }))
    } else {
      this.setState(()=>({
        error : ""
      }))
      this.props.onSubmit({
        description: this.state.description,
        createdAt: this.state.createdAt.valueOf(),
        amount: parseFloat(this.state.amount,10)*100,
        note: this.state.note,
      })
    }
  }
  onDateChange = (createdAt)=>{
    if(createdAt){
      this.setState(()=>({createdAt}))
    }
  }
  onFocusChange = ({focused}) => {
    this.setState(()=> ({calendarFocus: focused}))
  }
  handleInputChange = (e)=>{
    const input = e.target.name;
    const value = e.target.value;
    this.setState(()=>({
      [input]: value,
    }))
  }
  render() {
    return (
      <div>
        {this.state.error && <h4> {this.state.error} </h4>}
        <form onSubmit={this.handleSubmit}>
          <input type="text"
            placeholder="description"
            name="description"
            autoFocus
            onChange={this.handleInputChange}
            noValidate = {true}
            value = {this.state.description}
            
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocus} // PropTypes.bool
            onFocusChange={this.onFocusChange}
            id="your_unique_id" // PropTypes.string.isRequired,
            numberOfMonths={2}
            isOutsideRange = {()=> false}
          />
          <input type="number"
            name="amount"
            placeholder="Amount"
            onChange={this.handleInputChange}
            value = {this.state.amount}
          />
          <textarea type="text"
            name="note"
            placeholder="Notes"
            onChange={this.handleInputChange}
            value = {this.state.note}
          />
          <button> Add Expense </button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
