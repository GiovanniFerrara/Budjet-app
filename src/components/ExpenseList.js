import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses';
import { startFetchExpenses } from '../actions/expenses'

export class ExpenseList extends Component {
  componentDidMount() {
    this.props.fetchExpenses()
  }
  render() {
    return (
      <div>
        <h1>Expense List</h1>
        {this.props.expenses.length === 0 ?
          <p>Loading...</p> :
          this.props.expenses.map((item) => (
            <ExpenseListItem {...item} key={item.id} />
          ))}
      </div>
    )
  }
}
const mapDispatchToProps = (dispatch) => ({
  fetchExpenses: () => {
    dispatch(startFetchExpenses())
  }
})

const mapStateToProps = (state) => {
  return { expenses: getVisibleExpenses(state.expenses, state.filters) }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
