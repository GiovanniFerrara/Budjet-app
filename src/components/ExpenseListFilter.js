import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';

// text: '',
//   sortBy: 'date',
//   startDate: undefined,
//   endDate: undefined

const ExpenseListFilter = ({ filters, dispatch }) => {

  return (
    <div>
      <input type="text" value={filters.text} onChange={(e) => {
        dispatch(setTextFilter(e.target.value))
      }} />
      <select value={filters.sortBy} onChange={(e) => {
        e.target.value === "amount" && dispatch(sortByAmount());
        e.target.value === "date" && dispatch(sortByDate());
      }}>
        <option value="date">Date</option>
        <option value="amount"> Amount</option>
      </select>
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}

export default connect(mapStateToProps)(ExpenseListFilter);
