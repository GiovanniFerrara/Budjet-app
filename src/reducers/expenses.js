import { editExpense } from "../actions/expenses";

// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {

    case 'ADD_EXPENSE':
      return addExpense(state, action.expense);

    case 'REMOVE_EXPENSE':
      return removeExpense(state, action.id)

    case 'EDIT_EXPENSE':
      return editExpenseById(state, action.id, action.updates)

    case 'FETCH_EXPENSES':
      return setExpenses(action.expenses)

    default:
      return state;
  }
};

const setExpenses = (expenses) => {
  return expenses
}

const addExpense = (state, expenseObj) => {
  return [...state, expenseObj]
}

const removeExpense = (state, expenseId) => {
  return state.filter(({ id }) => id !== expenseId)
}

const editExpenseById = (state, id, expenseUpdatesObj) => {
  return state.map((expense) => {
    if (expense.id === id) {
      return {
        ...expense,
        ...expenseUpdatesObj
      };
    } else {
      return expense;
    };
  });
}