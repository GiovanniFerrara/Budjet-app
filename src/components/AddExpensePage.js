import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends Component {

  onSubmit = (expense) => {
    this.props.startAddExpense(expense)
    this.props.history.push("/dashboard")
  }
  render() {
    return (
      <div>
        This is from my add expense component
    <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => {
    dispatch(startAddExpense(expense))
  }
}
)

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
