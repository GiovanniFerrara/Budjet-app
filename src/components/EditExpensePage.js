import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses'

const EditExpensePage = ({dispatch,history,expense,match}) => (
  <div>
    This is from my edit expense component
    <ExpenseForm 
      onSubmit ={ (newExpense)=> {
        console.log(newExpense)
        dispatch(editExpense(expense.id,newExpense));
        history.push("/")
      }}
      expense = {expense}
    />
    
    <button onClick={()=>{
      dispatch(removeExpense({id: expense.id}))
      history.push("/")
    }
      }> Remove </button>
  </div>
);
const mapStateToProps = (state, props)=>({
  expense: state.expenses.find((expense)=>{
    return expense.id = props.match.params.id
  })
})
export default connect(mapStateToProps)(EditExpensePage)
