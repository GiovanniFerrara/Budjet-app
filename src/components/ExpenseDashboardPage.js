import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseList />
    <ExpenseListFilter />
    <button onClick={() => {
      console.log(process.env.FIREBASE_API_KEY)
    }}>click</button>
  </div>
);

export default ExpenseDashboardPage;
