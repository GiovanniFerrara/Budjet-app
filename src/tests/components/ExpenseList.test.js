import React, { Component } from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses';


test('it should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
})


test('it should render ExpenseList without expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})