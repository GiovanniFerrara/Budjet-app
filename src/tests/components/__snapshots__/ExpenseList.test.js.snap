// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note="bla ble bli"
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note="bla"
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note="blu ble blo"
  />
</div>
`;

exports[`it should render ExpenseList without expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses found
  </p>
</div>
`;
